### PR TITLE
[IMP] web_editor, *: enable to use image options after saving an image

### PR DESCRIPTION
**Pull request by loco-odoo**
[REM] web_editor: remove unused original_id parameter

This commit removes the `original_id` parameter of the declaration of
the `modify_image()` function as it is not used inside of the function
since [1].

[1]: https://github.com/odoo/odoo/commit/aa8d2a8be2c8f713b4b39b2aa1ea2aaf1839396c

task-3046425

---------------------------------------------------------------------------------------------------------------------------------------------------------

[IMP] base, *: enable to use image options after saving an image

*: product, web_editor

Before this commit, the options "Shape, "Filter", "Format", "Quality"
and "Transform" disappeared after replacing and saving any images stored
in the database. The goal of this commit is to enable the use of these
options even after saving a replaced image. To do so, the data related
to those options are stored in the "description" field of the attachment
linked to the image. This attachment can be recovered thanks to the
model name from where the image comes from, the id of the model and the
field name of the image in the model.

Here is the flow that happens at the save of an image:
- In `_saveModifiedImage()`, the data attributes and classes related to
the options are collected and the description that will be stored in the
attachment is build. It has the form
`| dataX:valueX ... classes:A,B,... |`. In this expression, `| ` and
` |` are only there to easily recover the data in the description of the
attachment (as this solution also works if the description of the
attachment is not originally empty).
- At the call of `save_embedded_field()` in the backend,
`_update_attachment_metadata()` is called leading to the update of the
description field of the attachment linked to the image. Note that this
is done after the `write` of the field. Indeed, it is only after this
step that the attachment linked to the image is created in the case of
a new product where its product image has been uploaded through the
media dialog.

At the load of an image (`loadImageInfo()`), the backend provides the
description of the attachment linked to the image. Thanks to it, the
data attributes of the image can be correctly set and the options
displays the correct values.

task-3046425

---------------------------------------------------------------------------------------------------------------------------------------------------------

[IMP] *: display the format of a saved image on the image options

*: web_editor

Steps to reproduce the problem:
- Replace a product image by one of your own (jpeg for example).
- Since [this commit], the mimetype of the image is webp. Replace the
format of the image by the original one.
- Save and edit.
- Click on the image to display its options.

-> Problem: "None" is display as the "Format".

The problem is that the `naturalWidth` of the image changes after the
save request. Let's say that it is 1281px at the upload through the
media dialog, it stays the same up to the reload of the iframe at the
save request. More investigation would be needed to know why but after
that, the `naturalWidth` of the image changes. This leads the "Format"
option to display `None`. Indeed, `_computeWidgetState()` returns the
new `naturalWidth` of the image and its mimetype. However, as it is not
listed by `_computeAvailableFormats()`, `None` is displayed for the
"Format".

[this commit]: https://github.com/odoo/odoo/commit/0449fe85cb0e1d639a4e1aeba26e90906f79254d

task-3046425

---------------------------------------------------------------------------------------------------------------------------------------------------------

[IMP] web_editor: pause the editor observer while loading an image

The goal of this commit is to pause the editor observer when loading the
image information. Indeed, without this commit, `.o_dirty` would be
added on the product image when selecting this image on the editor. This
was not the case in stable version because:
- If the image was not an embedded field (inside an `arch` for example
-> image on the website main page): the `loadImageInfo()` operation was
done at the drop of the image on the page (through
`_initializeImage()`). At the save of the `arch` field, the
`originalSrc` attribute was saved inside the field so the
`loadImageInfo()` process was not done the next time the user edited the
page and clicked on the image. In this case, `.o_dirty` was added
because there was a new image on the page.
- If the image was linked to an embedded field (`image_1920` of
`product.template` for example), the `loadImageInfo()` process was only
done if the user uploaded a new image (and so `.o_dirty` had to be
added). Indeed, in this case, the `/web_editor/get_image_info` rpc
returned a correct `original` leading to the complete `loadImageInfo()`
process flow. If the image was not a new uploaded image, the rpc was not
able to return the correct `original`.

Since one of the goals of this task is to adapt the
`/web_editor/get_image_info` rpc so that it returns the correct
attachment linked to a field of a record and because the `originalSrc`
attribute is not saved for embedded fields, the `loadImageInfo()` logic
will always apply when a user enters in edit mode and clicks on an
image linked to an embedded field. As we do not want the `o_dirty` class
to be added after this flow, the editor observer has to be paused when
loading the image information.

task-3046425

### CODE REVIEW

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
Here are my comments.
Note also that the branch name does not follow the guidelines:
- You made a typo `matser` => `master`. It is the reason you have a warning on the runbot.
- It should finish by `-loco`, it's easier for people to find your branch (on the runbot for example)

Do not recreate a PR because of this. It's just to let you know :+1:

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (loco-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Unfortunately, I don't see any alternative for the moment. Indeed, the line of code you propose could lead to unexpected behavior for some snippets such as the "carousel". Indeed, in this case, the querySelector will always return the first background image and the other one will not be accessible.

**Review by sobo-odoo — CHANGES_REQUESTED** [review]
Here is my review :slightly_smiling_face: 
It seems to work well. However, I observed some issues:
- when we change the image of the product and then go to the /shop page, the image is still the previous one. Is it normal or should that be fixed ?

Also, not related to what you did but maybe some things that you could fix while you are at it (ask before :wink:) :

- From 16.0, there is no way to remove a shape when we are in the shop because the cross to remove them is not there.
Maybe you could add it back (maybe it should be an other task and a fix from 16.0)? It is useful because since the shape is not reset when we replace or remove the image, it is impossible to remove it.

- The "Remove" button (in a product image) doesn't seem to work well: 1x/2 it doesn't remove anything or keep the shapes/filters/...

**Comment by loco-odoo** [comment]
> * when we change the image of the product and then go to the /shop page, the image is still the previous one. Is it normal or should that be fixed ?

It is now handle in this PR :+1:

 

> * From 16.0, there is no way to remove a shape when we are in the shop because the cross to remove them is not there.
>   Maybe you could add it back (maybe it should be an other task and a fix from 16.0)? It is useful because since the shape is not reset when we replace or remove the image, it is impossible to remove it.
> * The "Remove" button (in a product image) doesn't seem to work well: 1x/2 it doesn't remove anything or keep the shapes/filters/...

Two other tasks have been created for those problems :+1:

**Comment by loco-odoo** [comment]
Last review by Mehdi (mou)

**Comment by loco-odoo** [comment]
Last review by Benoit (bso)

**Comment by loco-odoo** [comment]
Note : the description field of the attachement is used to store data such as the `shape`, the `shapeColors`, the `filter`, the `resizeWidth` and the `quality` of the image.  For the moment, those information are recorded as text. However, we could also use JSON. The advantage of the JSON is that we do not need regex to recover the information but the advantage of the text is that the system can still work if there is a text description added by the user.

**Comment by qsm-odoo** [comment]
@robodoo delegate=bso-odoo

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (bso-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5935]: "apply on all type" => "applied on all types of" ?

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (bso-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:6505]: Is this not also needed in the first commit ?

**Review by bso-odoo — COMMENTED** [review]
Hi, here are a few comments and suggestions.
You will most likely have to rebase this PR onto webp after it is merged. :crossed_fingers: 
I also stored information in the `description` field of `ir.attachment` - so we will have to make sure both PRs do not destroy each other's data.
I think webp only stores data upon creation - and yours seems to mainting existing description - so it might actually already work fine. To be thoroughly tested.

File: addons/web_editor/controllers/main.py
Comment (loco-odoo) [inline addons/web_editor/controllers/main.py:561]: Note: before this commit, this was not needed as it was impossible to modify an image that came from the backend; the user had to first upload an image through the media dialog before being able to modify it.

**Comment by loco-odoo** [comment]
Note: waiting for https://github.com/odoo/odoo/pull/130329 to be merged

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js
Comment (loco-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1162]: Note : This will be done in this [PR] as it should be a fix in stable as well

[PR]: https://github.com/odoo/odoo/pull/130472

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js
Comment (loco-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:3046]: At this stage, we know that the image has been modified. If `description` is empty, the `oe-tmp-embedded` data attribute should still be added as it could mean that the user removed all the options (shape, filter, etc...) previously set. Thanks to the empty `oe-tmp-embedded`, the data related options will be removed from the db.

**Comment by loco-odoo** [comment]
@qsm-odoo , this is the task we spoke about last week if you want to have a look :) I let the fix of the second commit in order to be able to correctly test the feature but as soon as there is a fix for the saved absolute URL, we will be able to merge this fix from 16.4 :+1:

File: addons/web_editor/static/src/js/editor/image_processing.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/image_processing.js:451]: @loco-odoo Don't remember: is not that the thing we fully reverted in stable? Should not it be added back in stable (in its fixed version)?

File: addons/web_editor/static/src/js/editor/image_processing.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/image_processing.js:462]: Should not that be a fix in stable? (real question)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Quick first look

File: addons/web_editor/static/src/js/editor/image_processing.js
Comment (loco-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:451]: @qsm-odoo, This fix should be added back in stable but it can not be done now. Indeed, [this commit](https://github.com/odoo/odoo/commit/3528f2c3f9d3b30266f3421672986d202c724a76) fixed the problem of the absolute URL being stored after an image move, however, clients that did the move before the fix still have an absolute URL stored in their db. If I am not mistaken, @sobo-odoo is working on a task to convert the absolute URL to relative ones and it is after the merge of its task that this fix can be merged. Indeed, the goal of this fix is to perform the `loadImageInfo` operation for the images that do not have the `mimetypeBeforeConversion` attribute. To do so, the `src` argument has to be correct in order for the `/web_editor/get_image_info` rpc call to correctly return `original`.

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:451]: @loco-odoo We won't convert user URLs back to relative ones but make it so our code is more robust to work with absolute ones. In any case, this fix should be made in stable then forward-ported. If it has to wait for something before being merged, so be it.

@sobo-odoo Are you working on something like that? (make it so the code does work with absolute URLs). If not, decide which one of you is going to fix this :+1: Arguably, this could be made in the same PR, maybe even same commit as this stable fix about `mimetypeBeforeConversion`.

Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:451]: @qsm-odoo @loco-odoo I didn't start yet but I was planning to do it (it was lower in my priority list) :slightly_smiling_face: If it is urgent, I will try to do it this week :)

File: addons/web_editor/static/src/js/editor/image_processing.js
Comment (loco-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:462]: We could add it in stable in order to be more accurate but it will not fix a flow. Indeed:
- If the image is not an embedded field (it is inside an `arch` for exemple -> image on the website main page): the `loadImageInfo` operation is done at the drop of the image on the page (through `_initializeImage`). At the save of the `arch` field, the `originalSrc` attribute is saved inside the field so the `loadImageInfo` process will not be done the next time the user edit the page and click on the image. In this case, `.o_dirty` is added because there is a new image on the page.
- If the image is linked to an embedded field (`image_1920` of `product.template` for example), the `loadImageInfo` process will only be done if the user upload a new image (and so `.o_dirty` has to be added). Indeed, in this case, the `/web_editor/get_image_info` rpc will return a correct `original` leading to the complete `loadImageInfo`process flow. If the image is not a new uploaded image, the rpc will not be able to return the correct `original` (Note that it is one of the goal of this task to adapt the `/web_editor/get_image_info`  rpc so that it returns the correct attachment linked to a field of a record).

-> In those two cases, the `loadImageInfo` process is done only if the image is new so `.o_dirty` has to be added anyway.
However, this fix should be added when the fix of the second commit of this task will be merged in stable as a the `loadImageInfo` process will be done for images that are not necessarily new on the page.

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:462]: I see, thanks for clarifying :+1: Although, feels strange in the end to actually prevent the o_dirty class to be added for a DOM change we would actually like to be saved in the DOM (to not go through `get_image_info` again) but I understand. We'll get read of those dataset values at some point anyways, we'll re-debate at that point. For now this is good like this, for master :+1: Maybe just add this clarification in the commit message itself? :+1:

File: addons/web_editor/static/src/js/editor/image_processing.js
Comment (sobo-odoo) [reply addons/web_editor/static/src/js/editor/image_processing.js:451]: @loco-odoo it's done :wink: See [this PR](https://github.com/odoo/odoo/pull/136951) (for the master version of the fix)

**Comment by loco-odoo** [comment]
> If it has to wait for something before being merged, so be it.

@qsm-odoo , now that [this commit](https://github.com/odoo/odoo/commit/94b55d913b179b033bfc74f8fb1545675c75db9b) has been merged, this PR has everything it needs to be merged :)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Sorry for the delay checking this again. I did not do a full review, I have a main question that needs answering first :grimacing:

**Comment by loco-odoo** [comment]
Closed as it will be done in https://github.com/odoo/odoo/pull/155575

**Pull request closed without merging**
